StorageCipher initialization failed

Hi There,

First thanks for latest 3.3.4 update. Nice to see this is still under development and active.

Now issue I am facing is that on Android 9,I get following error : 

StorageCipher initialization failed
java.security.ProviderException: Failed to load generated key pair from keystore

Have tested on Androi 6,7,10. Issue occurs only on andoir 9/Pie.

I am simply using  :  
```
final storage = new FlutterSecureStorage();
await storage.read(key: 'sortedEncryptionKey');
```

Following are the logs  :  

```
D/FlutterSecureStoragePl(31078): Initializing StorageCipher
I/fluttersecurestorage(31078): Creating keys!
I/fluttersecurestorage(31078): Initializing
I/fluttersecurestorage(31078): Generating key pair
E/FlutterSecureStoragePl(31078): StorageCipher initialization failed
E/FlutterSecureStoragePl(31078): java.security.ProviderException: Failed to load generated key pair from keystore
E/FlutterSecureStoragePl(31078): 	at android.security.keystore.AndroidKeyStoreKeyPairGeneratorSpi.loadKeystoreKeyPair(AndroidKeyStoreKeyPairGeneratorSpi.java:530)
E/FlutterSecureStoragePl(31078): 	at android.security.keystore.AndroidKeyStoreKeyPairGeneratorSpi.generateKeyPair(AndroidKeyStoreKeyPairGeneratorSpi.java:472)
E/FlutterSecureStoragePl(31078): 	at java.security.KeyPairGenerator$Delegate.generateKeyPair(KeyPairGenerator.java:727)
E/FlutterSecureStoragePl(31078): 	at com.it_nomads.fluttersecurestorage.ciphers.RSACipher18Implementation.createKeys(RSACipher18Implementation.java:189)
E/FlutterSecureStoragePl(31078): 	at com.it_nomads.fluttersecurestorage.ciphers.RSACipher18Implementation.createRSAKeysIfNeeded(RSACipher18Implementation.java:120)
E/FlutterSecureStoragePl(31078): 	at com.it_nomads.fluttersecurestorage.ciphers.RSACipher18Implementation.<init>(RSACipher18Implementation.java:38)
E/FlutterSecureStoragePl(31078): 	at com.it_nomads.fluttersecurestorage.ciphers.StorageCipher18Implementation.<init>(StorageCipher18Implementation.java:31)
E/FlutterSecureStoragePl(31078): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin.ensureInitStorageCipher(FlutterSecureStoragePlugin.java:75)
E/FlutterSecureStoragePl(31078): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin.access$000(FlutterSecureStoragePlugin.java:29)
E/FlutterSecureStoragePl(31078): 	at com.it_nomads.fluttersecurestorage.FlutterSecureStoragePlugin$MethodRunner.run(FlutterSecureStoragePlugin.java:183)
E/FlutterSecureStoragePl(31078): 	at java.lang.Thread.run(Thread.java:764)
E/FlutterSecureStoragePl(31078): Caused by: java.security.UnrecoverableKeyException: Failed to obtain information about key
E/FlutterSecureStoragePl(31078): 	at android.security.keystore.AndroidKeyStoreProvider.getKeyCharacteristics(AndroidKeyStoreProvider.java:234)
E/FlutterSecureStoragePl(31078): 	at android.security.keystore.AndroidKeyStoreProvider.loadAndroidKeyStoreKeyPairFromKeystore(AndroidKeyStoreProvider.java:300)
E/FlutterSecureStoragePl(31078): 	at android.security.keystore.AndroidKeyStoreKeyPairGeneratorSpi.loadKeystoreKeyPair(AndroidKeyStoreKeyPairGeneratorSpi.java:521)
E/FlutterSecureStoragePl(31078): 	... 10 more
E/FlutterSecureStoragePl(31078): Caused by: android.security.KeyStoreException: -22
E/FlutterSecureStoragePl(31078): 	at android.security.KeyStore.getKeyStoreException(KeyStore.java:839)
E/FlutterSecureStoragePl(31078): 	at android.security.keystore.AndroidKeyStoreProvider.getKeyCharacteristics(AndroidKeyStoreProvider.java:236)
E/FlutterSecureStoragePl(31078): 	... 12 more
```

I will be very thankful if i get some help here.

--- a/flutter_secure_storage/android/build.gradle
+++ b/flutter_secure_storage/android/build.gradle
@@ -13,9 +13,6 @@ buildscript {
 }
 
 rootProject.allprojects {
-    tasks.withType(JavaCompile) {
-        options.compilerArgs << "-Xlint:unchecked" << "-Xlint:deprecation"
-    }
     repositories {
         google()
         mavenCentral()

--- a/flutter_secure_storage/android/build.gradle
+++ b/flutter_secure_storage/android/build.gradle
@@ -8,11 +8,14 @@ buildscript {
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.3.1'
+        classpath 'com.android.tools.build:gradle:4.2.2'
     }
 }
 
 rootProject.allprojects {
+    tasks.withType(JavaCompile) {
+        options.compilerArgs << "-Xlint:unchecked" << "-Xlint:deprecation"
+    }
     repositories {
         google()
         mavenCentral()
@@ -23,7 +26,6 @@ apply plugin: 'com.android.library'
 
 android {
     compileSdkVersion 30
-    buildToolsVersion '28.0.3'
 
     defaultConfig {
         minSdkVersion 18

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStoragePlugin.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/FlutterSecureStoragePlugin.java
@@ -72,8 +72,10 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
     @SuppressWarnings("unchecked")
     private void ensureInitialized(Map<String, Object> arguments) {
         Map<String, Object> options = (Map<String, Object>) arguments.get("options");
+        boolean useEncryptedSharedPreferences = false;
         if (options != null) {
-            if(useEncryptedSharedPreferences(options) &&
+            useEncryptedSharedPreferences = useEncryptedSharedPreferences(options);
+            if(useEncryptedSharedPreferences &&
                     Build.VERSION.SDK_INT >= Build.VERSION_CODES.M){
                 if(!(preferences instanceof  EncryptedSharedPreferences)){
                     try {
@@ -85,7 +87,7 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
             }
         }
 
-        if (storageCipher == null) {
+        if (storageCipher == null && !useEncryptedSharedPreferences) {
             try {
                 storageCipher = new StorageCipher18Implementation(applicationContext);
             } catch (Exception e) {
@@ -170,12 +172,12 @@ public class FlutterSecureStoragePlugin implements MethodCallHandler, FlutterPlu
     }
 
     private void write(String key, String value, boolean useEncryptedSharedPreference) throws Exception {
-        byte[] result = storageCipher.encrypt(value.getBytes(charset));
         SharedPreferences.Editor editor = preferences.edit();
 
         if(useEncryptedSharedPreference){
             editor.putString(key, value);
         } else {
+            byte[] result = storageCipher.encrypt(value.getBytes(charset));
             editor.putString(key, Base64.encodeToString(result, 0));
         }
         editor.apply();

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/ciphers/RSACipher18Implementation.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/ciphers/RSACipher18Implementation.java
@@ -1,15 +1,11 @@
 package com.it_nomads.fluttersecurestorage.ciphers;
 
-import android.annotation.SuppressLint;
 import android.annotation.TargetApi;
 import android.content.Context;
 import android.content.res.Configuration;
-import android.content.res.Resources;
 import android.os.Build;
 import android.security.keystore.KeyGenParameterSpec;
 import android.security.keystore.KeyProperties;
-import android.security.keystore.StrongBoxUnavailableException;
-import android.util.Log;
 
 import java.math.BigInteger;
 import java.security.Key;
@@ -163,7 +159,6 @@ class RSACipher18Implementation {
                 .build();
     }
 
-    @SuppressLint("NewApi")
     private void createKeys(Context context) throws Exception {
         final Locale localeBeforeFakingEnglishLocale = Locale.getDefault();
         try {

--- a/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/ciphers/StorageCipher18Implementation.java
+++ b/flutter_secure_storage/android/src/main/java/com/it_nomads/fluttersecurestorage/ciphers/StorageCipher18Implementation.java
@@ -1,6 +1,5 @@
 package com.it_nomads.fluttersecurestorage.ciphers;
 
-import android.annotation.SuppressLint;
 import android.content.Context;
 import android.content.SharedPreferences;
 import android.util.Base64;
@@ -13,7 +12,6 @@ import javax.crypto.Cipher;
 import javax.crypto.spec.IvParameterSpec;
 import javax.crypto.spec.SecretKeySpec;
 
-@SuppressLint("ApplySharedPref")
 public class StorageCipher18Implementation implements StorageCipher {
 
     private static final int ivSize = 16;
@@ -45,7 +43,6 @@ public class StorageCipher18Implementation implements StorageCipher {
                 return;
             } catch (Exception e) {
                 Log.e("StorageCipher18Impl", "unwrap key failed", e);
-                encrypted = new byte[0];
             }
         }
 
@@ -55,7 +52,7 @@ public class StorageCipher18Implementation implements StorageCipher {
 
         byte[] encryptedKey = rsaCipher.wrap(secretKey);
         editor.putString(AES_PREFERENCES_KEY, Base64.encodeToString(encryptedKey, Base64.DEFAULT));
-        editor.commit();
+        editor.apply();
     }
 
     @Override
@@ -100,12 +97,12 @@ public class StorageCipher18Implementation implements StorageCipher {
 
         SharedPreferences.Editor oldEditor = oldPreferences.edit();
         oldEditor.remove(AES_PREFERENCES_KEY);
-        oldEditor.commit();
+        oldEditor.apply();
 
         SharedPreferences newPreferences = context.getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE);
         SharedPreferences.Editor newEditor = newPreferences.edit();
         newEditor.putString(AES_PREFERENCES_KEY, existedSecretKey);
-        newEditor.commit();
+        newEditor.apply();
     }
 
 }

--- a/flutter_secure_storage/example/android/app/build.gradle
+++ b/flutter_secure_storage/example/android/app/build.gradle
@@ -16,25 +16,22 @@ apply plugin: 'com.android.application'
 apply from: "$flutterRoot/packages/flutter_tools/gradle/flutter.gradle"
 
 android {
-    compileSdkVersion 28
-    buildToolsVersion '28.0.3'
+    compileSdkVersion 30
 
     lintOptions {
         disable 'InvalidPackage'
     }
 
     defaultConfig {
-        // TODO: Specify your own unique Application ID (https://developer.android.com/studio/build/application-id.html).
         applicationId "com.it_nomads.fluttersecurestorageexample"
         minSdkVersion 18
-        targetSdkVersion 28
+        targetSdkVersion 30
         versionCode 1
         versionName "1.0"
     }
 
     buildTypes {
         release {
-            // TODO: Add your own signing config for the release build.
             // Signing with the debug keys for now, so `flutter run --release` works.
             signingConfig signingConfigs.debug
         }

--- a/flutter_secure_storage/example/android/build.gradle
+++ b/flutter_secure_storage/example/android/build.gradle
@@ -1,18 +1,18 @@
 buildscript {
     repositories {
         google()
-        jcenter()
+        mavenCentral()
     }
 
     dependencies {
-        classpath 'com.android.tools.build:gradle:3.3.2'
+        classpath 'com.android.tools.build:gradle:4.2.2'
     }
 }
 
 allprojects {
     repositories {
         google()
-        jcenter()
+        mavenCentral()
     }
 }
 

--- a/flutter_secure_storage/example/android/gradle.properties
+++ b/flutter_secure_storage/example/android/gradle.properties
@@ -1,4 +1,3 @@
 org.gradle.jvmargs=-Xmx1536M
 android.enableJetifier=true
 android.useAndroidX=true
-android.enableR8=true

--- a/flutter_secure_storage/example/android/gradle/wrapper/gradle-wrapper.properties
+++ b/flutter_secure_storage/example/android/gradle/wrapper/gradle-wrapper.properties
@@ -1,6 +1,6 @@
-#Thu Feb 14 23:07:34 SGT 2019
+#Mon Aug 02 13:25:04 CEST 2021
 distributionBase=GRADLE_USER_HOME
+distributionUrl=https\://services.gradle.org/distributions/gradle-7.1.1-bin.zip
 distributionPath=wrapper/dists
-zipStoreBase=GRADLE_USER_HOME
 zipStorePath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions/gradle-4.10.1-all.zip
+zipStoreBase=GRADLE_USER_HOME

